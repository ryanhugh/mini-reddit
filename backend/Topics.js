

// Model to keep track of the topics. 


class Topics {

  constructor(props) {

    // There are a couple different ways that it is possible to keep track of these topics. 

    // It would be possible to optimize for either direct topic access and store the topics in a hash map or optimize for sorting times and keep the topics in a list.
    //      This would make vote changing O(1) because we know where every item is, but would have to sort the topics ever time they are retrieved from the server. 

    // It would also be possible to keep the items sorted in memory. This would mean that the topics would not have to be sorted every time they are processed.
    //      However, it would mean that changing any information about topics is more complicated. This would require a linear scan of all the objects to find the one that the user is editing. 
    //      Then, once it is found, we may have to change the topic's location in the array if the score of the topic was changed. 

    // It would also be possible to have multiple data structures referencing the same objects. This would be complicated, however. 

    // I chose to stick with a simple array of topics to avoid over-engineering this problem. 
    // More optimizations could be done if we are expecting a very large number of topics or if the requirements change. 


    // Array to keep track of all the topics. 
    this.topics = [];
  }

  addTopic(text) {

    this.topics.push({
      text: text,
      score: 0,
      id: this.topics.length
    })
  }

  // Increase the score of a topic. 
  upvote(id) {
   if (!this.topics[id]) {
      return;
    }

    this.topics[id].score ++;
  }

  // Decrease the score of a topic. 
  downvote(id) {
   if (!this.topics[id]) {
      return;
    }

    this.topics[id].score --;
  }

  getTopics() {
    return this.topics;
  }

}


export default Topics;