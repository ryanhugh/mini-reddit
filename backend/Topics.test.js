import Topics from './Topics';


it('should be able to add a topic', function() {
	
	let instance = new Topics();
	expect(instance.getTopics().length).toBe(0);
	instance.addTopic('Hi there.')

	expect(instance.getTopics().length).toBe(1);

	instance.addTopic('Hi there again.')

	expect(instance.getTopics().length).toBe(2);

});


it('should be able to upvote a topic', function() {
	
	let instance = new Topics();
	instance.addTopic('Hi there.')
	
	let topicId = instance.getTopics()[0].id

	instance.upvote(topicId)

	expect(instance.getTopics()[0].score).toBe(1);

});


it('should be able to downvote and upvote a topic', function() {


	let instance = new Topics();
	instance.addTopic('Hi there.')
	
	let topicId = instance.getTopics()[0].id

	instance.upvote(topicId)
	instance.upvote(topicId)
	instance.downvote(topicId)
	instance.downvote(topicId)
	instance.downvote(topicId)

	expect(instance.getTopics()[0].score).toBe(-1);
});


it('should be able to downvote and upvote a topic', function() {


	let instance = new Topics();
	instance.addTopic('Hi there.')
	instance.addTopic('Hi there again.')
	
	let firstTopicId = instance.getTopics()[0].id
	let secondTopicId = instance.getTopics()[0].id

	instance.upvote(firstTopicId)
	instance.upvote(firstTopicId)
	instance.downvote(secondTopicId)
	instance.downvote(secondTopicId)
	instance.downvote(secondTopicId)

	expect(instance.getTopics()).toMatchSnapshot();
});