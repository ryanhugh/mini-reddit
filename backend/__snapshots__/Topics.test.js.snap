// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to downvote and upvote a topic 1`] = `
Array [
  Object {
    "id": 0,
    "score": -1,
    "text": "Hi there.",
  },
  Object {
    "id": 1,
    "score": 0,
    "text": "Hi there again.",
  },
]
`;
